### ISSUE TITLE
Leaky Bucket implementation

### ISSUE THREAD

**Issue by blackprince001** [ikeasamoahansah/rate-x#1]
# Implement Leaky Bucket Algorithm for Rate Limiter

## Overview
Implement the Leaky Bucket algorithm as a concrete implementation of our rate limiting framework's `RateLimitingAlgorithm` base class. This will provide a queue-based rate limiting solution that processes requests at a consistent rate.

## Algorithm Description
The Leaky Bucket algorithm works by:
- Maintaining a fixed-size queue (bucket) of requests
- Processing requests at a constant rate (like water leaking from a hole)
- New requests are rejected when the bucket is full
- Follows FIFO (First In, First Out) principle for request processing

### Class Structure
```python
from rate_limiter.algorithms.base import RateLimitingAlgorithm
from rate_limiter.core.request import RateLimitRequest
from rate_limiter.core.response import RateLimitResponse

class LeakyBucketAlgorithm(RateLimitingAlgorithm):
    def __init__(self, bucket_size: int, leak_rate: float):
        """
        Initialize the leaky bucket algorithm
        
        Args:
            bucket_size: Maximum number of requests the bucket can hold
            leak_rate: Number of requests processed per second
        """
        pass
```

### Required Method Implementations

1. `allow_request` method:
   - Check if bucket has capacity
   - Handle concurrency safely
   - Return appropriate RateLimitResponse

2. `get_status` method:
   - Report current bucket fill level
   - Report processing statistics
   - Return algorithm configuration

### Technical Requirements

1. Performance
   - O(1) time complexity for allow/reject decisions
   - Efficient queue management

2. Metrics
   - Track request acceptance/rejection rates
   - Monitor bucket utilization
   - Record processing latency

3. HTTP Integration
   - Set appropriate rate limit headers
   - Include retry-after times when rejecting
   - Handle request metadata

## Testing Scenarios

1. Basic Functionality
   - Bucket fills correctly
   - Requests are processed at the specified rate
   - FIFO order is maintained

2. Edge Cases
   - Empty bucket behavior
   - Full bucket behavior
   - Concurrent request handling
   - Rate limit boundary conditions

3. Integration Tests
   - FastAPI middleware integration
   - Header handling
   - Request identification

## Acceptance Criteria

- [ ] Implements RateLimitingAlgorithm interface correctly
- [ ] Handles concurrent requests safely
- [ ] Maintains constant processing rate
- [ ] Sets appropriate HTTP headers
- [ ] Includes comprehensive test suite
- [ ] Documentation with usage examples
- [ ] Performance benchmarks

## Resources
- Project's algorithm base class: `src/rate_limiter/algorithms/base.py`
- Core request/response models: `src/rate_limiter/core/`
- [Leaky Bucket Algorithm Overview](https://en.wikipedia.org/wiki/Leaky_bucket)

**Comment by thesarfo** [comment]
#6

**Comment by ikeasamoahansah** [comment]
closed with #6
